growthbook/growthbook
# Add pro to licenses. New accounts on cloud use licenses.

### Features and Changes
- Update the UI according to: https://www.figma.com/file/azr2BhItAh86kgVYbVshlW/Account-Upgrade-Flows?type=design&node-id=15-2886&mode=design&t=BO05DZvYfMrG3Zet-0
- Made enterprise/license.ts not reliant on their being a single licenseData or cacheDate so that it can be used in cloud.
- Got rid of setLicense - the one place we used it we were worried that the org could get into a state of not having any license if it already had one and tried to put a new invalid one.  Since we now use caches for licenses and don’t overwrite the cache unless we are already done fetching/verifying the new license, we can’t get into this situation.
- Added a license controller and moved a few functions from admin controller there.
- Added new email verification flow for endpoints on self-hosted
- Moved all calls to the license server to enterprise (this way LICENSE_SERVER is only defined in one place, can’t be swapped out without breaking enterprise license, and the front end doesn’t need to connect to an outside ip address).  Front-end calls are now proxied through the back-end. 
- Moved the new pro trial endpoint out of the existing subscription/checkout, and have it directly call the stripe API, as it was confirmed that if the trial is free and will auto cancel we don’t legally need to go to stripe.  This simplifies the flow for the trial endpoint.
- Made the new subscription endpoints (including /success which is called after a new license is created) to use licenses.  The billing, and quote endpoints which are used to update existing stripe subscriptions continue to work for subscription data on organizations but have been updated to also handle licenses.  The webhook endpoint is left as is handling subscription data on organizations and will eventually be removed totally and handled on the license server.
- We now always call success/ refresh license when anything gets updated on stripe’s side thereby removing need of the web hooks except for expiring trial emails. 
- Add emails from invites as part of the license seats.
- Move license from /user response to /organization response as licenses live either site wide or on an organization, while users can have many organizations and hence many licenses.
- Continue to show the upgrade button when the current state is trials.
- For Current subscription on organization in cloud added a call to refresh the license data when coming back to billing (before this I think we relied on the web hook eventually updating - now it should show the data immediately).
- Added new CLOUD_SECRET that needs to match between cloud and license server as a way to verify that they are on cloud.  On cloud we automatically say their email is verified.
- Modified message on adding a new user on pro to make it clear they won’t be charged until their trial ends, and show it on self-hosted pro as well.
- Got rid of a duplicate test to make sure we return the LICENSE_KEY env var’s data for when licenseKey argument is not set.
- Added an additional test to make sure we don’t accidentally set the LICENSE_SERVER to the dev server
- Fixed a test to make sure the old style license keys never expired which was not testing what was expected of it.
- Added TODOs in a lot of places to get rid of code once we move all organizations to use licenses.

### Dependencies

- License Server PR needs to be landed and deployed
- CLOUD_SECRET set on License Server

### Testing
Change LICENSE_SERVER at the top of enterprise/src/license.ts to "http://localhost:8080/api/v1/";

Show License:
SH:
- Put a valid license on an org
- Try putting an invalid license on an org
- Refresh the page and see that the old one remains
- Update the license on the license server
- Click Refresh and see the new data
- Put a valid old style license
- Click Download Report
- Open downloaded file and see correct contents
Cloud:
- Make sure that on general settings you can’t edit the license key manually (no pencil icon)
- Update the license on the license server
- Click Refresh and see the new data

### Screenshots

- TODO: Will make Looms for all the flows
